### ISSUE TITLE
Consider using all threads on a single instance as the default setting

### ISSUE THREAD

**Issue by Eugenio-Bruno** [niklasf/fishnet#117]
Hi, I've conducted a small investigation of play quality and nodes per second as stockfish scales with the amount of threads, and I think a reasonable conclusion from this - probably after doing longer testing and on different kinds of hardware - is that we might want to use as a default setting a single process with all threads assigned to it.

I'm new to chess programming, so if my reasoning here is wrong, feel free to just shoot down the idea. It was just a couple of days worth of cycles.

All testing on a Ryzen 3700X. 

```
ResultSet-EloRating>los 0 5 9
              fish_2_c fish_4_c fish_1_c fish_8_c fish_16_
fish_2_core             5295428  6777443  9746332  9966953
fish_4_core    4704571           6501885  9695527  9958167
fish_1_core    3222556  3498114           9324612  9879446
fish_8_core     253667   304472   675387           7733188
fish_16_core     33046    41832   120553  2266811
```

For a fixed node count (4M, similar to the average I get when actually running fishnet), it seems that 1, 2, 4, and 8 cores are almost definitively better than 16 threads (hyperthreading), but probably very close to each other in quality of play.

```
Rank Name           Elo    +    - games score oppo. draws
   1 fish_2_core     17   25   25   399   54%    -4   71%
   2 fish_4_core     16   25   25   400   54%    -4   69%
   3 fish_1_core      9   25   25   399   52%    -2   70%
   4 fish_8_core    -15   25   25   398   46%     4   68%
   5 fish_16_core   -27   25   25   398   43%     7   71%
```

8 cores *seems* worse than <4 cores, but even if it is, it's by 20 elo... I'm inclined to say we shouldn't care about it, and the same goes for 16 threads. I don't think 30-40 elo of difference for this application is huge.

I thought that hyperthreading wouldn't add nps - I was wrong. It does, and a lot.

`timeout.exe 5m python -m fishnet run --cores 15 --threads-per-process 15`
`[fishnet v1.17.2] Analyzed 1136 positions, crunched 4421 million nodes`

`timeout.exe 5m python -m fishnet run --cores 8 --threads-per-process 8`
`[fishnet v1.17.2] Analyzed 882 positions, crunched 3392 million nodes`

I've done a 10 minute run for every value of --threads-per-process between 1 and 15 and found no big difference for throughput.

For latency though, the issue seems simpler. The more threads per process, proportionally less latency for new requests to be processed. Additionally, throughput doesn't greatly decrease.

I analyzed a 40 move game in 13 seconds on my home machine, which has worse specs than many gaming machines. When a game gets analyzed this fast, I think it is a huge quality of life improvement for an user.

`https://lichess.org/sHdTSEH3 took 13.4s (0.2s per position)`

**Comment by Eugenio-Bruno** [comment]
I look forward to your thoughts.

This was the quick and dirty test to grid search for throughput, which didn't show significant differences by just inspecting the logs:

```
import subprocess

for use_hyperthreading in [True, False]:
    if use_hyperthreading:
        cores = 15
    else:
        cores = 8

    for n_threads_per_process in range(1, cores + 1):
        run_name = "HT_" + ("on" if use_hyperthreading else "off") + "--" + "threads_per_process_" + str(n_threads_per_process)
        print("Starting", run_name)

        proc = subprocess.Popen(
            ["python", "-m", "fishnet", "run", "--cores", str(cores), "--threads-per-process", str(n_threads_per_process)],
            stdout = subprocess.PIPE, stderr = subprocess.PIPE
        )

        try:
            output, error = proc.communicate(timeout = 600)
        except subprocess.TimeoutExpired:
            proc.kill()
            output, error = proc.communicate()

        open("fishnet_test/" + run_name + ".stdout.txt", "w").write(output.decode("utf8"))
        open("fishnet_test/" + run_name + ".stderr.txt", "w").write(error.decode("utf8"))
```

This was the cutechess-cli script to do the tournament, which resulted in about 1000 games before I stopped it early:

```
for /f "tokens=2 delims==" %%a in ('wmic OS Get localdatetime /value') do set "dt=%%a"
set "YY=%dt:~2,2%" & set "YYYY=%dt:~0,4%" & set "MM=%dt:~4,2%" & set "DD=%dt:~6,2%"
set "HH=%dt:~8,2%" & set "Min=%dt:~10,2%" & set "Sec=%dt:~12,2%"

set "datestamp=%YYYY%%MM%%DD%" & set "timestamp=%HH%%Min%%Sec%"
set "fullstamp=%YYYY%-%MM%-%DD%_%HH%-%Min%-%Sec%"

cutechess-cli.exe -pgnout sf_core_count_a_%fullstamp%.pgn -repeat -games 2 -rounds 10000 -openings file=C:\Users\io\Desktop\noob_2moves.pgn ^
    -engine name=fish_1_core dir=E:\badgine\posgen\ cmd=stockfish_20070107_x64_modern.exe proto=uci option.Threads=1 nodes=4000000 tc=120+120 ^
    -engine name=fish_2_core dir=E:\badgine\posgen\ cmd=stockfish_20070107_x64_modern.exe proto=uci option.Threads=2 nodes=4000000 tc=120+120 ^
    -engine name=fish_4_core dir=E:\badgine\posgen\ cmd=stockfish_20070107_x64_modern.exe proto=uci option.Threads=4 nodes=4000000 tc=120+120 ^
    -engine name=fish_8_core dir=E:\badgine\posgen\ cmd=stockfish_20070107_x64_modern.exe proto=uci option.Threads=8 nodes=4000000 tc=120+120 ^
    -engine name=fish_16_core dir=E:\badgine\posgen\ cmd=stockfish_20070107_x64_modern.exe proto=uci option.Threads=16 nodes=4000000 tc=120+120
```

For simplicity sake, stockfish here is just downloaded from the stockfish releases and is the version mentioned there, it's not the one that comes with fishnet, but I don't think the conclusion can be wildly different.

I'm also going to attach the PGN resulted from the above script, which I then analyzed with bayeselo, if you want to do your own analysis. GitHub seems to require me to zip it.

[sf_core_count_a_2020-07-10_11-48-15.zip](https://github.com/niklasf/fishnet/files/4907352/sf_core_count_a_2020-07-10_11-48-15.zip)

**Comment by niklasf** [comment]
Thanks for running these experiments. Two questions:

Which hashtable size were the engines using here? (I don't think it matters much, but just for the record).

Should we still draw the line somewhere? Occasionally we have contributors with ridiculous machines upwards of 64 cores.

**Comment by Eugenio-Bruno** [comment]
> Which hashtable size were the engines using here? (I don't think it matters much, but just for the record).

Whoops, that might actually be a bit of a problem with my analysis. I'm using the default stockfish hash size, which might be quite small. Should I rerun the experiment with another hashsize?

>Should we still draw the line somewhere? Occasionally we have contributors with ridiculous machines upwards of 64 cores.

We see performance deteriorating at 16 threads, but not enough to be concerning. Going up to 64 might start to be bad, however, though I don't own a machine with that many cores to run that test.

**Comment by Eugenio-Bruno** [comment]
There is another issue. I found this in the stockfish discord:

```
Score of stockfish_20061707_1t vs stockfish_20061707_12t: 364 - 82 - 554 [0.641]
Elo difference: 100.7 +/- 14.2, LOS: 100.0 %, DrawRatio: 55.4 %

1000 of 1000 games finished.
1 thread vs 12 threads, fixed 1 million nodes.
```

This doesn't specify what machine it is or what UCI options are used (though I'll ask for these details), but this seems to imply double the difference that I found.

It might be hardware related, or maybe have something to do with it being 1M nodes instead of 4M, or the hash size... but whatever it is, it makes the decision not quite as clear.
[hert_op._sf_fix_node_1m_1tv12t.pgn.zip](https://github.com/niklasf/fishnet/files/4937959/hert_op._sf_fix_node_1m_1tv12t.pgn.zip)

**Comment by jim-perkins** [comment]
The one thing that strikes me with the 24 cores that I am currently donating to lichess:  During Non Peak times threads sit idle where they could be utilized to provide a faster response for a real person waiting on an analysis.  

Currently have threads at the default setting as requested here.
https://github.com/niklasf/fishnet/commit/0a77a7cb22c789f653d581cc0fb9605adca4705a#comments

**Comment by Eugenio-Bruno** [comment]
That's a great point, @jim-perkins. At that time, it doesn't really matter whether throughput* is 10-20% lower, it only matters how quickly users can get their analysis from the (generously, thank you!) donated hardware.


* After some discussion in the stockfish subreddit, I believe the amount of nps grows linearly with the amount of cores, it's just playing strength that doesn't.

**Comment by jim-perkins** [comment]
That's true Eugenio.  The other side is that for non user analysis '  AKA cheat analysis '  giving up 10-20% throughput does add up.  Initially upon donating some cores I conducted a very crude test.  After an analysis completed on my server,  clicked the link to view the analysis and looked to see if a live person was waiting for the results.  The 10-20% does matter hence I accepted niklasf request to 'leave it at the default'.  Maybe it will up to programmers to decide if breaking out the 'live' requests for additional cores is worth the effort.

**Comment by jim-perkins** [comment]
dont keep history of CPU usage but do have power consumption which is pretty much the same thing on this server.  Would be fair to say this is rough graph of fishnet utilization of my server.  Time zone: America/Chicago (CDT, -0500)  (min,avg,max)
![image](https://user-images.githubusercontent.com/6559269/87814816-0d90da80-c82a-11ea-8738-ec50781219c2.png)

**Comment by Eugenio-Bruno** [comment]
@gekkenhunter from the stockfish discord has genereously helped me with this. The more nodes, the less the elo difference. The concerning test above was at 1M nodes. At a fixed 4M nodes per position, which is similar to what fishnet uses, the difference between *one* and *twelve* threads is just 50 elo. Considering our application is finding blunders and such, and correlating top engine moves with cheater moves, I don't think 50 elo less than the best engine on the planet at 12 threads vs 1 core is that bad.

```
[Event "1v12threads. Hash =1024 mb. Tc= 4M nodes per move"]
Elo difference is 48 +/- 10
````

We also need to consider throughput of course - assuming fixed nodes, increasing cores per process doesn't lose significant strength (I think) - but does it decrease amount of nodes crunched?

I've done a very limited benchmark on my 4c/8t machine. Running fishnet with --threads-per-process 8 --cores 8, I get 6.9 mnps, while with --threads-per-process 1 --cores 8, I get 6.0 mnps. Yes, I get *more* nps with a single process than with less processes. Tomorrow I'll run a grid search on my 3700X workstation, sweeping from 1 thread per process to 15 and from 1 to 16 (with --cores 15, the default, and --cores 16).

In my opinion, this is a great tradeoff. Consider how much time chess.com analysis take (and, well, how long analysis takes right now on lichess). I think it's a *great* user experience to be able to have an analysis in seconds instead of a couple of minutes. Think about people playing 3+0, where the analysis latency might be a significant portion of the time they spend on the website.

**Comment by bftjoe** [comment]
Don't you also have to increase hash size or this change is pretty pointless?

**Comment by Eugenio-Bruno** [comment]
Didn't seem like hash size was the determining factor in fishnet performance, at least according to my benchmarks, but if you have different data/tests that shows otherwise, they're very welcome! I'm sure the maintainers will keep making default settings better as we come to different conclusions.

**Comment by bftjoe** [comment]
Hash size matters to the quality of the analysis. As you have more threads
you _need_ a biggher hash table. Not sure you're even measuring move
quality when all you're doing is fixed number of nodes analysis.

**Comment by Eugenio-Bruno** [comment]
I might be misunderstanding, but fishnet does 3.5M nodes per position, and so I asked someone on the SF discord to try to check how much worse the quality of analysis is with a varying amount of threads, and this was the result. Note that the hash table between the 1T and 12T configurations are the same:

```
[Event "1v12threads. Hash =1024 mb. Tc= 4M nodes per move"]
Elo difference is 48 +/- 10
```

And so the change was based on the assumption that 50 elo worse is a good tradeoff for a much lower latency.

**Comment by bftjoe** [comment]
The point is that you could have almost no elo loss if you increase hash size as you increase threads. You're thrashing the hash table by making it too small.

**Comment by Eugenio-Bruno** [comment]
Ah, sorry for misunderstanding! I could set up a test on cutechess-cli but I'm probably going to mess it up. Any chance you could provide a command line for me or run it yourself with the default hash size 256MB and let's say three times as much and see if we regain engine strength that way?

**Comment by ornicar** [comment]
Just FYI fishnet does 4 meganodes per position. Mentioning it in case it's relevant.
